packages/adapter/src/lib/adapter/translationManager.ts

```
@@ -0,0 +1,99 @@
+    public async init(): Promise<void> {
+        const sysConfObj = await this.objects.getObject('system.config');
+
+        await this.updateLanguage(sysConfObj?.common.language || DEFAULT_LANG);
+    }
+
+    /**
+     * Set the language and load the translations
+     *
+     * @param lang the language to load the translations for
+     */
+    public async updateLanguage(lang: ioBroker.Languages): Promise<void> {
+        if (lang === this.loadedLanguage) {
+            return;
+        }
+
+        this.translations.clear();
+
+        this.loadedLanguage = lang;
+        let allTranslations: Record<string, string> = {};
+
+        for (const configDir of this.configDirs) {
+            let content: string;
+
+            try {
+                content = await fs.readFile(path.join(configDir, `${lang}.json`), { encoding: 'utf8' });
+            } catch (e) {
+                if (e.code !== 'ENOENT' || lang === DEFAULT_LANG) {
+                    throw e;
+                }
+
+                // fallback to default language if no translations are available for configured language
+                content = await fs.readFile(path.join(configDir, `${DEFAULT_LANG}.json`), { encoding: 'utf8' });
+            }
+            const newTranslations: Record<string, string> = JSON.parse(content);
+
+            allTranslations = { ...allTranslations, ...newTranslations };
+        }
+
+        this.translations = new Map(Object.entries(allTranslations));
+    }
+
+    /**
+     * Translate given key to the active language (configured in `system.config` object)
+     *
+     * @param key key to translate
+     * @param placeholders A dictionary of keys which will be replaced by their corresponding value
+     */
+    public translate(key: string, placeholders?: Record<string, string>): string {
+        let translation = this.translations.get(key);
+
+        if (translation === undefined) {
+            throw new Error(`No translation found for key "${key}"`);
```

Better return the key as it is usually done in the UIs? Throwing errors might be problematic

---

The rule should be
Return translatio
Return english trabnsation if no matching translation exists
Return key if no translation exusts at all.

An exception should not be raised

---

The translations have fallbacks when loaded, either provide full translations or no. It is a new method it does not replace FE translation methods in any way and keeping it clean is preferred from my side. A fail fast approach lets the developer know in time that used methods are wrongly applied.

---

In my oppinion an exception should NEVER be raised here. Its likely that some places exist in code which are not really executed during tests. Crashing the adapter at such places would be inappropiate. So every trying to avoid such situations would / should encapsolate the translate call into one central try/catch routine to dismiss the exeptiond and return the key. This could be done centrally too.

If a feedback is desired adapter code could log a warning that no translation for key xxx is available. This would be a feedback to dev without crashing an adapter.